proposals/NNNN-swiftpm-extensible-build-tools.md

```diff
@@ -0,0 +1,807 @@
+
+Providing even basic support for extensibility is expected to allow more codebases to be built using the Swift Package Manager, and to automate many steps that package authors currently have to do manually (such as generate source code manually and commit it to their package).
+
+## Proposed solution
+
+This proposal introduces a new target type called `packageExtension`.
+
+Package extensions are Swift targets that use specialized API in a new `PackageExtension` library (provided by SwiftPM) to create and configure commands to run during the build.
+
+The initial `PackageExtension` API described in this proposal is minimal and is mainly focused on source code generation, but this API is expected to grow over time to support new package extension capabilities.
+
+Package extensions are somewhat analogous to package manifests:  both are Swift scripts that are evaluated in sandboxed environments and use specialized APIs for a limited and specific purpose.  In the case of a package manifest, the purpose is to define those characteristics of a package that cannot be determined from the file system.  In the case of a package extension, it is to procedurally define new commands and dependencies to run during the build, based on a target structure and a set of build parameters.
+
+A package extension is invoked during the build system’s planning process, and is given access to the configuration of the target to which it applies. The package extension also has read-only access to the package directory of the target, and is also allowed to write to specially designated areas of the build output directory.
+
+Note that the extension itself does *not* perform the actual work of the build tool — that’s done by the command line invocation that the package extension creates.
+
+In this initial proposal, the package extension itself is not invoked during the build — however, the build commands that were created by the extension *are* run as part of any build in which they need to run, as determined by their input and output dependencies.
+
+This initial proposal provides only limited ways for a package target to configure the build tool extensions it uses. However, because extension have read-only access to the package directory, they can read custom configuration files as needed. While this means that configuration of the extension resides outside of the package manifest, it does allow each package extension to provide a configuration format suitable for its own needs. We see this pattern commonly used in practice already, in the form of configuration files that are used to configure source generators, etc. Future proposals are expected to let package extensions define options that can be controlled in the client package's manifest.
```

Comment:
(… because extension***s*** have …)

```suggestion
This initial proposal provides only limited ways for a package target to configure the build tool extensions it uses. However, because extensions have read-only access to the package directory, they can read custom configuration files as needed. While this means that configuration of the extension resides outside of the package manifest, it does allow each package extension to provide a configuration format suitable for its own needs. We see this pattern commonly used in practice already, in the form of configuration files that are used to configure source generators, etc. Future proposals are expected to let package extensions define options that can be controlled in the client package's manifest.
```